File: app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardReferrerStatsRowHolderViewModel.java
package com.kickstarter.viewmodels;


import android.support.annotation.NonNull;
import android.util.Pair;

import com.kickstarter.libs.ActivityViewModel;
import com.kickstarter.libs.Environment;
import com.kickstarter.libs.utils.NumberUtils;
import com.kickstarter.libs.utils.PairUtils;
import com.kickstarter.models.Project;
import com.kickstarter.services.apiresponses.ProjectStatsEnvelope;
import com.kickstarter.ui.viewholders.CreatorDashboardReferrerStatsViewHolder;

import rx.Observable;
import rx.subjects.PublishSubject;

public interface CreatorDashboardReferrerStatsRowHolderViewModel {
  interface Inputs {
    /** Current project and list of referrer stats. */
    void projectAndReferrerStatsInput(Pair<Project, ProjectStatsEnvelope.ReferrerStats> projectAndReferrerStats);
  }

  interface Outputs {
    /** Emits project and the amount pledged for this referrer. */
    Observable<Pair<Project, Float>> projectAndPledgedForReferrer();

    /** Emits string number of backers. */
    Observable<String> referrerBackerCount();

    /** Emits resource ID of referrer color. */
    Observable<Integer> referrerSourceColorId();

    /** Emits source name of referrer. */
    Observable<String> referrerSourceName();
  }

  final class ViewModel extends ActivityViewModel<CreatorDashboardReferrerStatsViewHolder> implements Inputs, Outputs {

    public ViewModel(final @NonNull Environment environment) {
      super(environment);

      this.projectAndPledgedForReferrer = this.projectAndReferrerStats
        .map(pr -> Pair.create(pr.first, pr.second.pledged()));

      this.referrerSourceColorId = this.projectAndReferrerStats
        .map(PairUtils::second)
        .map(referrerStat -> ProjectStatsEnvelope.ReferrerStats.referrerTypeEnum(referrerStat.referrerType()).getReferrerColorId());

      this.referrerSourceName = this.projectAndReferrerStats
        .map(PairUtils::second)
        .map(ProjectStatsEnvelope.ReferrerStats::referrerName);

      this.referrerBackerCount = this.projectAndReferrerStats
        .map(PairUtils::second)
        .map(ProjectStatsEnvelope.ReferrerStats::backersCount)
        .map(NumberUtils::format);
    }
    
    public final Inputs inputs = this;
    public final Outputs outputs = this;

    private final PublishSubject<Pair<Project, ProjectStatsEnvelope.ReferrerStats>> projectAndReferrerStats = PublishSubject.create();

    private final Observable<Pair<Project, Float>> projectAndPledgedForReferrer;
    private final Observable<String> referrerBackerCount;
    private final Observable<Integer> referrerSourceColorId;
    private final Observable<String> referrerSourceName;

    @Override public void projectAndReferrerStatsInput(final @NonNull Pair<Project, ProjectStatsEnvelope.ReferrerStats> projectAndReferrerStats) {
      this.projectAndReferrerStats.onNext(projectAndReferrerStats);
    }

    @Override public @NonNull Observable<Pair<Project, Float>> projectAndPledgedForReferrer() {
      return this.projectAndPledgedForReferrer;
    }
    @Override public @NonNull Observable<String> referrerBackerCount() {
      return this.referrerBackerCount;
    }
    @Override public @NonNull Observable<Integer> referrerSourceColorId() {
      return this.referrerSourceColorId;
    }
    @Override public @NonNull Observable<String> referrerSourceName() {
      return this.referrerSourceName;
    }
  }
}

File: app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardReferrerStatsRowHolderViewModel.kt
package com.kickstarter.viewmodels

import android.util.Pair
import com.kickstarter.libs.ActivityViewModel
import com.kickstarter.libs.Environment
import com.kickstarter.libs.utils.NumberUtils
import com.kickstarter.libs.utils.ObjectUtils
import com.kickstarter.libs.utils.PairUtils
import com.kickstarter.models.Project
import com.kickstarter.services.apiresponses.ProjectStatsEnvelope.ReferrerStats
import com.kickstarter.services.apiresponses.ProjectStatsEnvelope.ReferrerStats.Companion.referrerTypeEnum
import com.kickstarter.ui.viewholders.CreatorDashboardReferrerStatsViewHolder
import rx.Observable
import rx.subjects.PublishSubject

interface CreatorDashboardReferrerStatsRowHolderViewModel {
    interface Inputs {
        /** Current project and list of referrer stats.  */
        fun projectAndReferrerStatsInput(projectAndReferrerStats: Pair<Project, ReferrerStats>)
    }

    interface Outputs {
        /** Emits project and the amount pledged for this referrer.  */
        fun projectAndPledgedForReferrer(): Observable<Pair<Project, Float>>

        /** Emits string number of backers.  */
        fun referrerBackerCount(): Observable<String>

        /** Emits resource ID of referrer color.  */
        fun referrerSourceColorId(): Observable<Int>

        /** Emits source name of referrer.  */
        fun referrerSourceName(): Observable<String>
    }

    class ViewModel(environment: Environment) :
        ActivityViewModel<CreatorDashboardReferrerStatsViewHolder?>(environment), Inputs, Outputs {

        val inputs: Inputs = this
        val outputs: Outputs = this
        private val projectAndReferrerStats = PublishSubject.create<Pair<Project, ReferrerStats>>()
        private val projectAndPledgedForReferrer: Observable<Pair<Project, Float>>
        private val referrerBackerCount: Observable<String>
        private val referrerSourceColorId: Observable<Int>
        private val referrerSourceName: Observable<String>

        init {
            projectAndPledgedForReferrer = projectAndReferrerStats
                .map {
                    Pair.create(
                        it.first,
                        it.second.pledged()
                    )
                }

            referrerSourceColorId = projectAndReferrerStats
                .map { PairUtils.second(it) }
                .map { referrerStat: ReferrerStats -> referrerTypeEnum(referrerStat.referrerType()).referrerColorId }

            referrerSourceName = projectAndReferrerStats
                .map { PairUtils.second(it) }
                .map { it.referrerName() }

            referrerBackerCount = projectAndReferrerStats
                .map { PairUtils.second(it) }
                .map { it.backersCount() }
                .filter { ObjectUtils.isNotNull(it) }
                .map { requireNotNull(it) }
                .map {
                    NumberUtils.format(it)
                }
        }

        override fun projectAndReferrerStatsInput(projectAndReferrerStats: Pair<Project, ReferrerStats>) {
            this.projectAndReferrerStats.onNext(projectAndReferrerStats)
        }

        override fun projectAndPledgedForReferrer(): Observable<Pair<Project, Float>> =
            projectAndPledgedForReferrer

        override fun referrerBackerCount(): Observable<String> =
            referrerBackerCount

        override fun referrerSourceColorId(): Observable<Int> =
            referrerSourceColorId

        override fun referrerSourceName(): Observable<String> =
            referrerSourceName
    }
}
